### DISCUSSION TITLE
Content negotiation in Inertia responses

### DISCUSSION THREAD

**Discussion by SimoneStefani** [inertiajs/inertia#1114] [Help]
In the current Inertia implementation and according to the [protocol documentation page](https://inertiajs.com/the-protocol), Inertia requests (i.e. requests with the `X-Inertia: true` header) have the `Accept` header set to `text/html, application/xhtml+xml` while the expected response has `Content-Type: application/json`.

To my understanding this breaks the purpose of content negotiation. In fact certain frameworks (e.g. [Ktor](https://ktor.io/)) don't accept this scenario and throw `406 Not Acceptable`.

Any thoughts around this topic?

**Comment by ajnsn** [comment]
Hey @SimoneStefani 

Yes, Inertia is a rebel. I think the original reason why Inertia masquerades as an HTML request is that Jonathan wanted Inertia to always trigger the backend behavior for a normal HTML request, e.g. Laravel has this `unauthenticated` method inside the exception [Handler](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Foundation/Exceptions/Handler.php#L408) which would Inertia requests not redirect to the login page and instead provide a json response.

As you are totally right with your understanding and this would also be mine but I think content negogiation at it's heart is broken. It's getting ignored at all the time. The server can somehow  decide what it's willing to send, e.g. client requesting GIF and server responds with an PDF download. 

The [spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html) also seems to be very vague in terms of the 406 response. It's more of a "should" than a "must"?

There is also a wiki page at [WHATWG](https://wiki.whatwg.org/wiki/Why_not_conneg) that shows you all the downsides. This page is also linked fairly prominently on the [MDN page](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation).

**Comment by RobertBoes** [comment]
This is really interesting, this could also be the root cause of #1013 since Chrome might cache the wrong response because it was requested as an ordinary HTML page

**Comment by SimoneStefani** [comment]
Thanks for the responses! Totally understand where Inertia comes from and the state of content negotiation. Will look for a workaround for my specific use case.
